recap/src/Imp/SLang.hs

```diff
@@ -19,23 +19,53 @@ data GEQ a b where
   GEQS :: GEQ a b -> GEQ (S a) (S b)
 
 geqTrans :: GEQ a b -> GEQ b c -> GEQ a c
-geqTrans = undefined
+geqTrans _ GEQZ = GEQZ
+geqTrans (GEQS ab) (GEQS cd) = GEQS (geqTrans ab cd)
 
 class Numm a where
   geqRefl :: GEQ a a
 
 instance Numm Z where
-  geqRefl = undefined
+  geqRefl = GEQZ
 
 instance (Numm a) => Numm (S a) where
-  geqRefl = undefined
+  geqRefl = GEQS geqRefl
 
 data SafeExpr t i where
   Var ::
     GEQ i (S iv) ->
     iv ->
     SafeExpr Int i
--- ...
+  Const :: Int -> SafeExpr Int i
+  BinOp ::
+    GEQ i li ->
+    GEQ i ri ->
+    L.Op ->
+    SafeExpr Int li ->
+    SafeExpr Int ri ->
+    SafeExpr Int i
+  Read :: SafeExpr Int i
+  Write :: SafeExpr Int i -> SafeExpr () i
+  Seq ::
+    GEQ i li ->
+    GEQ i ri ->
+    SafeExpr () li ->
+    SafeExpr t ri ->
+    SafeExpr t i
+  If ::
+    GEQ i ci ->
+    GEQ i ti ->
+    GEQ i ei ->
+    SafeExpr Int ci ->
+    SafeExpr a ti ->
+    SafeExpr a ei ->
+    SafeExpr a i
+  LetIn ::
+    GEQ i ei ->
+    SafeExpr Int ei ->
+    SafeExpr a (S i) ->
+    SafeExpr a i
+  Skip :: SafeExpr () i
```

Comment:
Why do you need a `GEQ` for each sub-expression?

Reply:
Well my idea was to allow constructing expressions from subexpressions with another `i` parameter.
I guess with type inference it is unnecessary. I deleted this.